src/conv/heuristics/ai_heuristics.cpp

```diff
@@ -0,0 +1,445 @@
+    if(miopen::IsLogging(LoggingLevel::Info2))
+    {
+        std::stringstream ss;
+        for(auto& id : sol)
+            ss << solver::Id{id}.ToString() << " ID:" << id << ", ";
+        MIOPEN_LOG_I2("Heuristic Result: " << ss.str());
+    }
+    return sol;
+}
+} // namespace immed_mode
+#endif // MIOPEN_ENABLE_AI_IMMED_MODE_FALLBACK
+
+#if MIOPEN_ENABLE_AI_KERNEL_TUNING
+namespace tuning {
+
+Metadata::Metadata(const std::string& arch, const std::string& solver)
+{
+    const nlohmann::json metadata =
+        common::LoadJSON(GetSystemDbPath() + "/" + arch + "_" + solver + "_metadata.ktn.model");
+    num_tuning_params = metadata["num_tuning_params"].get<std::size_t>();
+    tuning_decodings = metadata["decodings"]["tunings"].get<std::unordered_map<std::string, int>>();
+}
+
+class Model
+{
+public:
+    Metadata metadata;
+    Model(const std::string& arch, const std::string& solver)
+        : metadata(Metadata(arch, solver)),
+          encoder(fdeep::load_model(EncoderPath(arch, solver), true, fdeep::dev_null_logger)),
+          decoder(fdeep::load_model(DecoderPath(arch, solver), true, fdeep::dev_null_logger))
+    {
+    }
+    virtual ~Model() = default;
+    fdeep::tensors Encode(std::vector<float> features, std::size_t dim) const
+    {
+        fdeep::tensor input_tensor = fdeep::tensor(fdeep::tensor_shape(dim, dim), features);
+        return encoder.predict({input_tensor});
+    }
+    fdeep::tensors Decode(const float prev_token, const fdeep::tensors& context) const
+    {
+        return decoder.predict(
+            {{fdeep::tensor(fdeep::tensor_shape(1), std::vector<float>(1, prev_token)),
+              context[0],
+              context[1],
+              context[2],
+              context[3]}});
+    }
+
+private:
+    const fdeep::model encoder;
+    const fdeep::model decoder;
+    static std::string EncoderPath(const std::string& arch, const std::string& solver)
+    {
+        return GetSystemDbPath() + "/" + arch + "_" + solver + "_encoder.ktn.model";
+    }
+    static std::string DecoderPath(const std::string& arch, const std::string& solver)
+    {
+        return GetSystemDbPath() + "/" + arch + "_" + solver + "_decoder.ktn.model";
+    }
+};
+
+std::shared_ptr<Model> GetModel(const std::string& arch, const std::string& solver)
+{
+    static const std::string prevArch{arch};
+
+    if(prevArch != arch)
+        MIOPEN_THROW("Cannot use AI tuning models for multiple gpu architectures");
+    static std::map<std::string, std::shared_ptr<Model>> models;
+    auto it = models.find(solver);
+    if(it == models.end())
+    {
+        std::shared_ptr<Model> model = std::make_shared<Model>(arch, solver);
+        models[solver]               = model;
+        return model;
+    }
+    else
+    {
+        return it->second;
+    }
+}
+
+bool ModelSetParams(const std::string& arch,
+                    const std::string& solver,
+                    std::vector<float> features,
```

Comment:
```suggestion
                    const std::vector<float>& features,
```

Reply:
Had to add variable (features) initalization/declaration within ModelRunParameters() to make this run, since this was initially passed as a temporary value (rvalue I believe) since previously I was passing the return value of the TransformFeatures function directly in the ModelSetParameters call.